### ISSUE TITLE
Feature: Batch Activities Browser with barcode scanner.

### ISSUE THREAD

**Issue by meteos77** [textbrowser/biblioteq#240]
The borrowing functionality through the barcode scanner works great. **Bravo for this evolution that I find major.**

Of course I have cases not yet taken into account :-) ;  namely books without isbn10 and without isbn13.

A barcode is assigned to these documents (without isbn) but it is in the accession_number field.

Can you include in Scan Identifier Ready, the search on the accession_number field ?

**Comment by textbrowser** [comment]
This is an odd request because any unique value can identify an item and therefore the term Identifier becomes lost.

**Comment by meteos77** [comment]
I don't find my request so strange from my point of view, accession number is a unique reference number, which is interesting as soon as the books are older than the isbn standard (30% of our library collection still).

https://en.wikipedia.org/wiki/Accession_number_(cultural_property)

**Comment by meteos77** [comment]
text suggestion : 
Scan Identifiers Ready (isbn , accession_number)

**Comment by textbrowser** [comment]
Why not also Alternate Identifier? OID. MYOID. Call Number. These are unique. The problem with such requests is that they create hidden variables. One has to look at the source to understand the identifier in question. What if you're looking for an ISBN and the Accession Number is also present in another item? Which item should I retrieve?

**Comment by textbrowser** [comment]
SELECT title FROM book WHERE accession_number = 'A' OR id = 'A' retrieves multiple items if they share 'A'.

**Comment by textbrowser** [comment]
This problem does not have a logical solution. It has a solution, but not in the realm of sets.

**Comment by textbrowser** [comment]
So you'd retrieve the one you may not want and you'd have to modify your results. You wouldn't know if it's the wrong number because of hidden variables. And if you knew somehow, you couldn't correct the query. And even if this was correctable, the INSERT statement would require a similar implementation because of the new field.

It gets uglier and uglier.

**Comment by textbrowser** [comment]
So. If it works 90% or 80% or 70%, that's a super majority. If there are weird instances, manual entry. Complicating everything for a small percent is not a good solution. Those are exceptions which are fine. Exceptions require other logic. Separate and free of hidden variables.

**Comment by meteos77** [comment]
if you take your washing machine and it doesn't wash your clothes 1 time out of 3 it's direction the repair.
if you take your car and it works 2 times out of 3 it's direction the repair.

for me 30% which does not function is not a majority ; but a concern nevertheless.

**Comment by meteos77** [comment]
as usual, I'm not a developer so if it's not ideal don't do it.

**Comment by meteos77** [comment]
in the context of BiblioteQ (book management)

the software should be "aware" that 
- if the number provided starts with 978 or 979
- that it is composed of 13 digits
- the last one is a precise calculation
-> this one is to be searched in the field isbn13.

for the others it is necessarily accession_number.

**Comment by meteos77** [comment]
`SELECT title FROM book WHERE accession_number = 'A' OR id = 'A' retrieves multiple items if they share 'A'.`

from my point of view, and in my base those are the exceptions

**Comment by meteos77** [comment]
you have triggered an idea :-)
possible improvement for the 70% -> Show title for a physical book check with scanned isbn.

**Comment by textbrowser** [comment]
For hidden variables, adding the entire content of the item does nothing. Why? Because you cannot query the correct item. The correct item doesn't exist.

Yo, I have two books here. Same title. Which one do you want? Mountain of Beautiful Men. OK. Here. That's not the one I want. Yes, it is. No, it isn't.

That's a problem without a resolution.

**Comment by textbrowser** [comment]
This is considered complete. I'm not adding a query which is meaningless and a subsequent modification query which is even more meaningless. A book is ISBN-unique. If it cannot be, that's another query. Not a just-in-case query.

**Comment by textbrowser** [comment]
If you want another query, submit a request for an enhancement. :P

**Comment by meteos77** [comment]
I find it a shame to have closed this ticket.
Librarians can spend 30-40% of their duty time on this feature; it should have been improved.

**Comment by textbrowser** [comment]
It's completed for books.

A universal uniqueness does not exist at the circulation level. ISBNs, accession numbers, barcodes, alternate identifiers, titles, etc. ISBNs have 10 or 13 characters. Accession numbers are limitless as are all of the other fields. When an item is scanned and ten entries are retrieved, that's not a solution. It's like scanning an item at the grocery and having 10 possible prices and then identifying the exact mango.

Practical and functional solutions are better than perceived perfection. Sure, it works in your case but doesn't generally work because it's illogical.

Hence, closed.

**Comment by textbrowser** [comment]
Some CDs have UPCs. Video games, DVDs have UPCs, perhaps. I'm not querying a 3rd party to determine the type. Unless someone provides a universal application to identify a video game or a DVD or a CD, I cannot guess the type via length. Journals have 8-digit ISSNs as do magazines. Batch Activities is not a search function. The information you provide it must be accurate. Transforming it into yet-another-search mechanism voids all of the other search functions. It's meant as a simple and correct mechanism and not as a solution for incorrect data management.

**Comment by textbrowser** [comment]
Grey Literature is also interesting because it's not necessarily properly identified and now it's allowed in Batch Activities. So it takes thought and work, what doesn't?

**Comment by meteos77** [comment]
I didn't want to come back to isbn and accession_number; I know you choose the best technically but to have display improvements for example

the user will type the accession number: m363742
It will be good that it is a return displayed to know at least his name and surname.
the librarian can ask the name to a person; not his number and in case of typing mistake, no verification is possible.
So we will assign documents to another member.

for the articles it is also the case, BQ can display the title of the document and the librarian makes a visual verification the displayed title and the document.
the title will take 1 second while checking the isbn is impossible.
this allows to check title <-> isbn at each output of a document; 
the base of a library may have been built without the barcode reader so entry errors may have occurred.

**Comment by meteos77** [comment]
after today's update, it sends me back
"Critical error! Missing table item(s), when using Batch Activities

another effect with the addition of the various categories (visible on the french)
it adds 2 clicks per book :-(
because book is in first position in English thus that does not change anything but "livre" (book) in French is in 5 position thus one must go in the pop-up specified book instead of CD.

**Comment by textbrowser** [comment]
> I didn't want to come back to isbn and accession_number; I know you choose the best technically but to have display improvements for example
> 
> the user will type the accession number: m363742 It will be good that it is a return displayed to know at least his name and surname. the librarian can ask the name to a person; not his number and in case of typing mistake, no verification is possible. So we will assign documents to another member.
> 
> for the articles it is also the case, BQ can display the title of the document and the librarian makes a visual verification the displayed title and the document. the title will take 1 second while checking the isbn is impossible. this allows to check title <-> isbn at each output of a document; the base of a library may have been built without the barcode reader so entry errors may have occurred.

This is duplicating richness which exists elsewhere. Why? In a functional setting, the item you're scanning is unique. For a book with an ISBN having multiple copies, you will scan the ISBN and the barcode identifying the item. Why add a title? A name? The item is in your hands. It must be the correct one. Why is Batch Activities responsible for validating the data-entry process. Batch Activities is a circulation process.

**Comment by textbrowser** [comment]
> after today's update, it sends me back "Critical error! Missing table item(s), when using Batch Activities
> 
> another effect with the addition of the various categories (visible on the french) it adds 2 clicks per book :-( because book is in first position in English thus that does not change anything but "livre" (book) in French is in 5 position thus one must go in the pop-up specified book instead of CD.

The more that is added, the more that is likely to fail.

**Comment by textbrowser** [comment]
This is not fun for me. Done for today and this month!

**Comment by textbrowser** [comment]
![Screenshot_20221015_132721](https://user-images.githubusercontent.com/10701156/195999896-29853593-72b3-4bf9-a285-83565fd4aa60.png)

**Comment by meteos77** [comment]
thank you very much for all the additions/changes

I don't ask you anything else until December

good continuation in your activities

**Comment by meteos77** [comment]
Thanks again, I appreciate your explanatory approach.

**Comment by meteos77** [comment]
last thing : `I've already violated this principle with hidden queries`

I'm not in favor of you going against the principles, so if this feature is a problem of principle, better remove it because it will cause problems later.

We will do it differently later or not.

see you soon ... in December :-)

**Comment by textbrowser** [comment]
> I didn't want to come back to isbn and accession_number; I know you choose the best technically but to have display improvements for example
> 
> the user will type the accession number: m363742 It will be good that it is a return displayed to know at least his name and surname. the librarian can ask the name to a person; not his number and in case of typing mistake, no verification is possible. So we will assign documents to another member.
> 
> for the articles it is also the case, BQ can display the title of the document and the librarian makes a visual verification the displayed title and the document. the title will take 1 second while checking the isbn is impossible. this allows to check title <-> isbn at each output of a document; the base of a library may have been built without the barcode reader so entry errors may have occurred.

Patron's do not have accession numbers. Patrons have identifiers. An identifier may be lastname_firstname or a value provided by BQ or something established by the library. If I introduce yet another verification mechanism in Batch Activities, it's yet another intermediate step to verify a value. If the value is incorrect, the circulation person has to access Members Browser. So, another intermediate verification process. OK. That requires a new query and two widgets. The reset mechanism needs revisions.

In short, BA is actually a verify-all process. Verify that the patron is correct. Verify the scanned value represents the physical item held by the librarian, verify everything. OK.

**Comment by textbrowser** [comment]
Which is again just weird. You requested a scanner-friendly interface. If I'm scanning a patron's card and items that I have in my hands, why do I need to verify the scanned data? If I do not have a barcode or a patron's ID or I am unsure of their ID, I find the ID in Members Browser. I copy that ID and paste it into Batch Activities. For items that I am unsure of because I do not have their barcodes, I search for them reserve them with the existing mechanism. The expectation that BA be scanner friendly while also discovering items that do not have scanner-friendly information is just... what?

But OK!

**Comment by textbrowser** [comment]
So, BA should verify the patron. Part A. Either verify them by their card or by some likeliness of the ID they tell the librarian. That is, if there are multiple members it should present them and the librarian locate the expected one. Step A. Turn BA into a copy of Members Browser. Step A: scanner-friendly while duplicating the identification of a member.

Step B. Retain scanner friendliness while duplicating the discovery process for items that do not have a scanning label.

Step C. Turn BA into a corrective process. If a member's data is incorrect, allow the member's information to be corrected.

Step D. If the member does not exist, provide the entry process in BA.

Step E. Correct the item through BA.

**Comment by textbrowser** [comment]
In a normal or abnormal library.

1. May I have your library card? 2. I don't have it. OK, what's your name and address? 3. Here it is.
2. Here are the items I'm reserving.
3. OK, let me scan them.

In more elaborate libraries, you perform that on your own using the items and your card.

**Comment by textbrowser** [comment]
The librarian doesn't ask me to remember my ID or force me to go home to get my card or have me verify my card with the name on file. Neither do they verify that the item in hand is the one scanned in. That's not the point of a reservation process. The patron is not quality control!

**Comment by meteos77** [comment]
I think I understood our difference.

See the following discussions:
Loan / Return #127
Batch Management #170

For me what you are developing is the answer to #127 and indeed for my vision, it necessarily takes what already exists elsewhere in the interface to make it fully usable with a single interface to apply the loan and return of documents via this tool.

If it is to answer #170 indeed, I understand all your remarks on why I ask all these improvements. And telling me I'm an idiot fits well :-)

**Comment by meteos77** [comment]
The previous case is the best at grandpa's.

if the user doesn't have his number or his card
that we go back to the procedure out of scanner and quite logical and I did not ask for evolution in these cases.
Going to the list of members and taking advantage of the filter field seems to me quite suitable.

**Comment by meteos77** [comment]
I had in mind that the functionality will be an extra-ordinary evolution.

I don't know if it's a mountain to develop because for me, there is really not much missing and the features exist in member.
For you it's a duplication while for the user it's doing a necessary action in the easiest way.

At the time of validation of the member BQ answers the name (this is used to verify that the entry is good when the member gives his number). effectively if card with barcode it becomes almost useless (except user-friendly).

have the icons
- list the borrowed items of the member
- "list the member's delays".
since we have provided the BA interface with a member, we might as well take advantage of updating the display of his loans in the main window without having to go through the member manager.

**Comment by meteos77** [comment]
when I see the main window
actions are available in the text menu and the same actions are available in the bar with icons.
for borrowing: right mouse button, icon bar and icon bar in the member manager 

I think it's great to be able to perform an action depending on the work area where you are

**Comment by meteos77** [comment]
example of improvement

BQ answers the NAME REPLY in the box at the validation of the member's ID

and in my understanding the icons on the right are
1 - shortcut to update in the main window the list of loans of the member specified in the member ID

2 - shortcut to update in the main window the list of overdue items of the member specified in member ID

3 - shortcut to open the member manager in case of difficulties

[biblioteq_batch_activities_browser.ui.tar.gz](https://github.com/textbrowser/biblioteq/files/9850545/biblioteq_batch_activities_browser.ui.tar.gz)

**Comment by textbrowser** [comment]
> if it's for #127
> 
> For the member part: indeed, we do not distribute cards, so no barcode for the moment . Currently with papy90s 1 - go to the borrowing interface 2 - enter member number (3 digits) + carriage return or scan his card
> 
> papy90s shows the name of the borrower and his borrowing history. In 2 actions : we know if the member has returned all the borrowed documents or if he has some left. this allows to make the point with the member.
> 
> in BQ 1 - go to the batch management interface 2 - enter member number (+ carriage return) or scan his card 3 - go to the member manager window 4 - click on the member 5 - click on "List items of the selected member".

Why do I need to see the patron's reserved items? If the maximum-reserved-books is set, the member cannot reserve additional items.

![Screenshot_20221024_102855](https://user-images.githubusercontent.com/10701156/197550776-36fc618f-5ac0-4795-aca9-1740ccc231cf.png)

**Comment by meteos77** [comment]
This request is not related to the number of books that can still be borrowed but some members forget to return books.

The librarian takes stock of the books that the member still has at home and that he/she must return to the library at his/her next visit.

Not all of our books stay in the library,
300 to 500 books per semester "roll over" with the central library.

**Comment by meteos77** [comment]
Attention we have to return the "title" book to the central library next month, remember to return it before.

**Comment by meteos77** [comment]
the system of exchange of books in large quantities is a peculiarity but it is the very interest of the library and of the relationship with the central library.

It allows very small libraries (like ours) to have a lot of books on loan. We can go up to 1000 loans at the same time.

With BiblioteQ, we can even increase the offer because it allows us to manage DVDs, CDs.

The central library even offers musical scores.
no category in BQ :-)

**Comment by textbrowser** [comment]
I see. This note about having to return an item is accounted for by the duration of a loan. See All Overdue. I can add a brief icon somewhere in BA which would indicate that the patron has reserved items which must be returned. The outstanding items are available for all and individual members.

**Comment by textbrowser** [comment]
> when I see the main window actions are available in the text menu and the same actions are available in the bar with icons. for borrowing: right mouse button, icon bar and icon bar in the member manager
> 
> I think it's great to be able to perform an action depending on the work area where you are

This depends on the context. You will notice that you view a member's overdue items through the Members Browser. The items are not displayed in the Members Browser, they're displayed in the main window. This is not duplication, instead, it creates a meaningful relationship. Such things can be added to Batch Activities. But again, duplicating the entire behavior of the main window in a compact interface is writing BQ within itself. The main window has SQL-engine specifics, states, progress dialogs, cancellation potential, etc.

**Comment by meteos77** [comment]
if it doesn't require too much development shortcuts will be perfect.
the interest of positioning them in BA is that we have already specified the member.

the information is of course displayed in the main window

**Comment by meteos77** [comment]
people borrow about 10-15 books maximum; maybe 30 for families with several children but this is rare.

We prefer to see members more often than to let them borrow too much material and not circulate it among members.

**Comment by meteos77** [comment]
good news because if the meaningful relationship icons are present, I don't see anything more to add to this feature :-)

**Comment by meteos77** [comment]
I think it's the natural evolution.
BQ celebrated its 20th anniversary. The handheld was much more expensive at the time of creation so the user interface was turned to the cheap device namely the mouse.

Now that barcode scanners are affordable, all libraries are almost equipped if the borrowing/return traffic is important.

BA will become the preferred interface for borrowing books.

**Comment by textbrowser** [comment]
Completed.
![Screenshot_20221024_112038](https://user-images.githubusercontent.com/10701156/197563512-3a8983f2-5c88-4db2-a00f-98171549f56c.png)

**Comment by textbrowser** [comment]
And you probably want also the overdue items too, right? :P

**Comment by meteos77** [comment]
of course it makes it easier to use :-)

**Comment by meteos77** [comment]
You have reached the ideal use of BQ/BA

1 - scan the limb or enter keyboard
2 - icons loan list or overdue list
allows to check the name of the member in the main window
allows you to check the status of the loans with the member
3 - actions recording loans

**Comment by textbrowser** [comment]
![Screenshot_20221024_120926](https://user-images.githubusercontent.com/10701156/197573753-f121b68c-977b-42de-8c7c-16d499b09ff6.png)

Notice the category: All Reserved.

**Comment by textbrowser** [comment]
So you don't need an additional button as the display will notify you of overdue items.

**Comment by textbrowser** [comment]
Format strings.

**Comment by meteos77** [comment]
the programming is really strict.

Indeed you are right, no need of the button the color allows to distinguish well the delays.

**Comment by meteos77** [comment]
I think that you have managed to have a simple function for the user and that meets the expectations of borrowing documents by hand shower.
BRAVO

**Comment by meteos77** [comment]
In the member field, can you activate the carriage return as validation?

**Comment by textbrowser** [comment]
> In the member field, can you activate the carriage return as validation?

There's nothing to validate. Pressing return or scanning a patron card will now activate something.

**Comment by textbrowser** [comment]
Nor should it. That's why there's a button for a specific task (like listing the reserved items).

**Comment by textbrowser** [comment]
The <scan> field is special because it has to work for scanners and return-pressed.

**Comment by meteos77** [comment]
I expected it to go to the field by itself "please scan an item"

**Comment by meteos77** [comment]
In any case very beautiful evolution today ... I was dreaming about it textbrowser does it :-)

**Comment by meteos77** [comment]
thank you for all the great modifications of the day !

**Comment by meteos77** [comment]
Welcoming the member then scanning the documents seems to me to be the ideal order, I don't understand the change?

**Comment by textbrowser** [comment]
> Welcoming the member then scanning the documents seems to me to be the ideal order, I don't understand the change?

I input the member once and scan multiple items.

**Comment by textbrowser** [comment]
1x Member.
Nx Scans.

**Comment by meteos77** [comment]
completely agree with this statement !

imagine the person has already borrowed too many books.
imagine the person is banned from borrowing as long as too many overdue books are on his account.

it's good to know before scanning 10 new items to know the status of the person's account.

after the change, we have to click on 
the member field and then the borrowing list button then return to scan id

before the change
button borrowing list then scan id

we lose 1 click :-(

**Comment by textbrowser** [comment]
So we're arguing about a focus now?

**Comment by textbrowser** [comment]
The title of the task is Borrow. What's the primary purpose of Borrow? Is it to view what someone has reserved? If that's the primary purpose, then Borrow is not borrow. The secondary benefits you requested are secondary.

Anyway, this is getting silly.

**Comment by textbrowser** [comment]
Actually, it's beyond silly. It's micromanagement. :P

**Comment by textbrowser** [comment]
You can take BQ and I can help you set which input you'd like to focus on first. That fork doesn't have to be merged. It's your fork and you can build it for your environment.

**Comment by meteos77** [comment]
It may be silly to you but the goal of batch management is also to have as few clicks as possible.

**Comment by meteos77** [comment]
Very interesting, but you don't have to give me too much advice to succeed.

**Comment by textbrowser** [comment]
I've taken two opposing views: one the scanner and one the mouse and blended them together.

If you fork BQ, you'll be correct in your brain and you won't have to convince me. I don't have to convince you.

I stated my rationale and you think it's incorrect. OK. What should I do? Dilute my rationale for your personal preference?

**Comment by meteos77** [comment]
no, I'm interested in doing my own little modification but I'm afraid I'll have to ask too much of you to be successful.

**Comment by textbrowser** [comment]
When such details are required and which would make the configuration of the software a nightmare, it's best that you step away and consider that it's open source.

**Comment by meteos77** [comment]
I could always have the new stuff you develop?

**Comment by textbrowser** [comment]
There's only one solution and that's a configuration. That suits everyone. But that little option takes a day or two or three.

**Comment by meteos77** [comment]
I can always try this adventure

**Comment by meteos77** [comment]
do you have a first advice to document me?

**Comment by meteos77** [comment]
As you wish :-)

Another important activity with the scanner is the return of the documents.
